tests/paramgrill.c

```diff
@@ -632,131 +1391,1112 @@ int benchFiles(const char** fileNamesTable, int nbFiles)
+        ZSTD_CCtx* const ctx = ZSTD_createCCtx();
+        ZSTD_DCtx* const dctx = ZSTD_createDCtx();
+        ZSTD_compressionParameters const CParams = ZSTD_getCParams(cLevel, blockSize, dictBufferSize);
+        if(BMK_benchParam(&candidate, origBuff, benchedSize, fileSizes, (U32)nbFiles, ctx, dctx, CParams)) {
+            ZSTD_freeCCtx(ctx);
+            ZSTD_freeDCtx(dctx);
+            ret = 3;
+            goto _cleanUp;
+        }
+
+        target.cSpeed = (U32)candidate.cSpeed; //Maybe have a small bit of slack here, like x.99? 
+        BMK_printWinner(stdout, cLevel, candidate, CParams, benchedSize);
+
+        ZSTD_freeCCtx(ctx);
+        ZSTD_freeDCtx(dctx);
+    }
+
+    g_targetConstraints = target;
 
     /* bench */
     DISPLAY("\r%79s\r", "");
-    DISPLAY("optimizing for %s - limit speed %u MB/s \n", inFileName, targetSpeed);
-    targetSpeed *= 1000000;
+    if(nbFiles == 1) {
+        DISPLAY("optimizing for %s", fileNamesTable[0]);
+    } else {
+        DISPLAY("optimizing for %lu Files", (unsigned long)nbFiles);
+    }
+    if(target.cSpeed != 0) { DISPLAY(" - limit compression speed %u MB/s", target.cSpeed / 1000000); }
+    if(target.dSpeed != 0) { DISPLAY(" - limit decompression speed %u MB/s", target.dSpeed / 1000000); }
+    if(target.cMem != (U32)-1) { DISPLAY(" - limit memory %u MB", target.cMem / 1000000); }
+    DISPLAY("\n");
+    findClockGranularity();
+
     {   ZSTD_CCtx* const ctx = ZSTD_createCCtx();
         ZSTD_DCtx* const dctx = ZSTD_createDCtx();
         winnerInfo_t winner;
-        BMK_result_t candidate;
+        U32 varNew[NUM_PARAMS];
         const size_t blockSize = g_blockSize ? g_blockSize : benchedSize;
+        U32 const maxNbBlocks = (U32) ((benchedSize + (blockSize-1)) / blockSize) + 1;
+        const size_t maxCompressedSize = ZSTD_compressBound(benchedSize) + (maxNbBlocks * 1024);
+        void* compressedBuffer = malloc(maxCompressedSize);
 
         /* init */
         if (ctx==NULL) { DISPLAY("\n ZSTD_createCCtx error \n"); free(origBuff); return 14;}
+        if(compressedBuffer==NULL) { DISPLAY("\n Allocation Error \n"); free(origBuff); free(ctx); return 15; }
         memset(&winner, 0, sizeof(winner));
         winner.result.cSize = (size_t)(-1);
 
+
         /* find best solution from default params */
-        {   const int maxSeeds = g_noSeed ? 1 : ZSTD_maxCLevel();
-            int i;
-            for (i=1; i<=maxSeeds; i++) {
-                ZSTD_compressionParameters const CParams = ZSTD_getCParams(i, blockSize, 0);
-                BMK_benchParam(&candidate, origBuff, benchedSize, ctx, dctx, CParams);
-                if (candidate.cSpeed < (double)targetSpeed) {
-                    break;
+        {
+            /* strategy selection */
+            const int maxSeeds = g_noSeed ? 1 : ZSTD_maxCLevel();
+            DEBUGOUTPUT("Strategy Selection\n");
+            if(varLen == NUM_PARAMS && paramTarget.strategy == 0) { /* no variable based constraints */  
```

Comment:
Combine these two branches. Instead of doing random initialization in the `else`, start with `ZSTD_getCParams()` and then apply the fixed parameters.